### DISCUSSION TITLE
Eloquent Collection Synth ignores sort order

### DISCUSSION THREAD

**Discussion by tinyoverflow** [livewire/livewire#7498] [7. Bugs (archived)] [answered]
### Livewire version

v3.2.2

### Laravel version

v10.34.2

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

1. Create a new Livewire component
2. Fetch an **Eloquent Collection** of models from the database with a sort order that does not represent the default sort order of the model keys (eg. sort by another field alphabetically).
3. Assign it to a public property inside the Livewire component class.
4. Use it in the frontend, eg. iterate over it to visualize the order.
5. Trigger a rehydration of the component, eg. by clicking a button side the component which triggers an action.

The sort order of the **Eloquent Collection** is reset to the default sort order, as the dehydrator for the Eloquent Collection only relies on the model keys. The sort order is not persisted. This can be seen inside the wire:snapshot data, as the IDs are stored as a plain JavaScript array without any keys. Therefore the serverside order won't be guaranteed.

There is basically just one practical workaround for this: Store the models in a normal collection by calling the `->collect()` method on the eloquent collection. This, of course, comes with a lot of overhead compared to the eloquent collection.

### Repository or Wirebox URL

https://github.com/tinyoverflow/livewire-eloquent-collection-order

### Additional information

I think Livewire should actually store the order of elements inside the eloquent collection. A suggestion would be to call `values()` to reset the keys and then use the keys to keep track of the actual order of items, instead of simply using a "1D" array of `modelKeys`.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by joshhanley** [comment, marked as answer]
@tinyoverflow we don't recommend storing Eloquent Collections in public properties, instead you should use Computed Properties. See here for more details https://www.youtube.com/watch?v=esD6o8HYV5Q&t=407s

**Reply by tinyoverflow** [reply]
@joshhanley That talk is pretty valuable, thanks for linking!

Even the caching option alone makes it really obvious, that a computed property is the way to go in this case. 

Case closed, I guess :)

**Reply by joshhanley** [reply]
@tinyoverflow thanks! 😊 glad it was helpful!

**Reply by icanall10** [reply]
This is not a solution

**Reply by tinyoverflow** [reply]
@icanall10 it actually is, at least in my case. Why isn’t this a solution for you? What are you trying to achieve? You might want to create a new topic for this.

**Reply by icanall10** [reply]
@tinyoverflow  my components contain many child components. and if I need to use the collection in a child component, I am forced to define computed methods in each child component and duplicate the code. I can simply pass a public property as a parameter

**Reply by joshhanley** [reply]
@tinyoverflow and @icanall10 we've just fixed this issue in PR #7359. So should be available in the next release 🙂
